scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt

```diff
@@ -137,16 +141,49 @@ class RunCoverage(
         println("\nComputed Coverage Ratio is: $computedCoverageRatio")
         println("\nGenerated report at: $reportOutputPath\n")
       }
-    } ?: println("No coverage reports generated.")
-  }
 
-  private fun runCoverageForTarget(testTarget: String): CoverageReport {
-    return runBlocking {
-      CoverageRunner(rootDirectory, scriptBgDispatcher, commandExecutor)
-        .runWithCoverageAsync(testTarget.removeSuffix(".kt"))
-        .await()
+      println("COVERAGE ANALYSIS COMPLETED.")
     }
   }
+
+  private fun runCoverageForTarget(testTarget: String): Deferred<CoverageReport> {
+    return CoverageRunner(rootDirectory, scriptBgDispatcher, commandExecutor)
+      .runWithCoverageAsync(testTarget.removeSuffix(".kt"))
+  }
+}
+
+private fun calculateAggregateCoverageReport(
+  coverageReports: List<CoverageReport>
+): CoverageReport {
+  fun aggregateCoverage(coverages: List<Coverage>): Coverage {
+    return if (coverages.contains(Coverage.FULL)) Coverage.FULL
+    else Coverage.NONE
+  }
+
+  val allCoveredLines = coverageReports.flatMap { it.coveredLineList }
+
+  val groupedCoveredLines = allCoveredLines.groupBy { it.lineNumber }
+
+  val aggregatedCoveredLines = groupedCoveredLines.map { (lineNumber, coveredLines) ->
+    CoveredLine.newBuilder()
+      .setLineNumber(lineNumber)
+      .setCoverage(aggregateCoverage(coveredLines.map { it.coverage }))
+      .build()
+  }
```

Comment:
Is there a test for verifying the behavior when the same line is hit in multiple tests for the single source file?

Reply:
Actually no, I didn't even realize to include the scenario as we didn't really considered the count for coverage check. 

But realizing that the existing sample tests does hit the same line no 7 -> `a + b` twice in its execution.

```
assertEquals(AddNums.sumNumbers(0, 1), 1)
assertEquals(AddNums.sumNumbers(3, 4), 7
```

But printing out the data it doesn't actually represent the `DA: 7, <execution_count as (2)>` instead it just shows it as `DA: 7, 1`

So, I tried [adding few more cases](https://github.com/oppia/oppia-android/pull/5459/commits/800ccdac2d7fbc821fe13c8a0ef06f41106daef9#diff-5a779ea4a8aa2caf5249ced27465f76b8ac19bed5cf250475f6033dfad464908R74) to hit the same line few more times, but the results were unchanged and just showed DA: 7, 1; while not hitting it once does show it as 0.

Also to check I tried running with few more actual files / targets as well and all their coverage data didn't actually represent their execution count (they were just either 1 or 0).

<details>
<summary>Coverage Reports</summary>

```kotlin
SF:domain/src/main/java/org/oppia/android/domain/auth/FirebaseAuthWrapperImpl.kt
FN:19,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl$signInAnonymously$1::onSuccess (Lcom/google/firebase/auth/AuthResult;)V
FN:22,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl$signInAnonymously$2::onFailure (Ljava/lang/Exception;)V
FN:7,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl::<init> (Lorg/oppia/android/domain/auth/FirebaseAuthInstanceWrapper;)V
FN:12,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl::getCurrentUser ()Lorg/oppia/android/domain/auth/FirebaseUserWrapper;
FN:17,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl::signInAnonymously (Lkotlin/jvm/functions/Function0;Lkotlin/jvm/functions/Function1;)V
FNDA:0,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl$signInAnonymously$1::onSuccess (Lcom/google/firebase/auth/AuthResult;)V
FNDA:0,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl$signInAnonymously$2::onFailure (Ljava/lang/Exception;)V
FNDA:1,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl::<init> (Lorg/oppia/android/domain/auth/FirebaseAuthInstanceWrapper;)V
FNDA:1,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl::getCurrentUser ()Lorg/oppia/android/domain/auth/FirebaseUserWrapper;
FNDA:0,org/oppia/android/domain/auth/FirebaseAuthWrapperImpl::signInAnonymously (Lkotlin/jvm/functions/Function0;Lkotlin/jvm/functions/Function1;)V
FNF:5
FNH:2
BRDA:12,0,0,1
BRDA:12,0,1,0
BRDA:12,0,2,0
BRDA:12,0,3,0
BRDA:12,0,4,0
BRDA:12,0,5,0
BRDA:17,0,0,-
BRDA:17,0,1,-
BRDA:17,0,2,-
BRDA:17,0,3,-
BRDA:17,0,4,-
BRDA:17,0,5,-
BRDA:18,0,0,-
BRDA:18,0,1,-
BRF:14
BRH:1
DA:7,1
DA:8,1
DA:9,1
DA:12,1
DA:13,0
DA:14,1
DA:17,0
DA:18,0
DA:19,0
DA:20,0
DA:21,0
DA:22,0
DA:23,0
DA:24,0
DA:26,0
DA:27,0
LH:5
LF:16
end_of_record

------------------------

SF:utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt
FN:58,org/oppia/android/util/parser/math/MathModel$MathModelSignature$Companion::createSignature$utility_src_main_java_org_oppia_android_util_parser_math_math_latex_model_kt (Ljava/lang/String;FZ)Lorg/oppia/android/util/parser/math/MathModel$MathModelSignature;
FN:33,org/oppia/android/util/parser/math/MathModel$MathModelSignature::<init> (Ljava/lang/String;IZ)V
FN:35,org/oppia/android/util/parser/math/MathModel$MathModelSignature::getLineHeightHundredX ()I
FN:34,org/oppia/android/util/parser/math/MathModel$MathModelSignature::getRawLatex ()Ljava/lang/String;
FN:36,org/oppia/android/util/parser/math/MathModel$MathModelSignature::getUseInlineRendering ()Z
FN:41,org/oppia/android/util/parser/math/MathModel$MathModelSignature::updateDiskCacheKey (Ljava/security/MessageDigest;)V
FN:14,org/oppia/android/util/parser/math/MathModel::<init> (Ljava/lang/String;FZ)V
FN:16,org/oppia/android/util/parser/math/MathModel::getLineHeight ()F
FN:15,org/oppia/android/util/parser/math/MathModel::getRawLatex ()Ljava/lang/String;
FN:17,org/oppia/android/util/parser/math/MathModel::getUseInlineRendering ()Z
FN:21,org/oppia/android/util/parser/math/MathModel::toKeySignature ()Lorg/oppia/android/util/parser/math/MathModel$MathModelSignature;
FNDA:1,org/oppia/android/util/parser/math/MathModel$MathModelSignature$Companion::createSignature$utility_src_main_java_org_oppia_android_util_parser_math_math_latex_model_kt (Ljava/lang/String;FZ)Lorg/oppia/android/util/parser/math/MathModel$MathModelSignature;
FNDA:1,org/oppia/android/util/parser/math/MathModel$MathModelSignature::<init> (Ljava/lang/String;IZ)V
FNDA:1,org/oppia/android/util/parser/math/MathModel$MathModelSignature::getLineHeightHundredX ()I
FNDA:0,org/oppia/android/util/parser/math/MathModel$MathModelSignature::getRawLatex ()Ljava/lang/String;
FNDA:1,org/oppia/android/util/parser/math/MathModel$MathModelSignature::getUseInlineRendering ()Z
FNDA:1,org/oppia/android/util/parser/math/MathModel$MathModelSignature::updateDiskCacheKey (Ljava/security/MessageDigest;)V
FNDA:1,org/oppia/android/util/parser/math/MathModel::<init> (Ljava/lang/String;FZ)V
FNDA:0,org/oppia/android/util/parser/math/MathModel::getLineHeight ()F
FNDA:0,org/oppia/android/util/parser/math/MathModel::getRawLatex ()Ljava/lang/String;
FNDA:0,org/oppia/android/util/parser/math/MathModel::getUseInlineRendering ()Z
FNDA:1,org/oppia/android/util/parser/math/MathModel::toKeySignature ()Lorg/oppia/android/util/parser/math/MathModel$MathModelSignature;
FNF:11
FNH:7
BRDA:46,0,0,1
BRDA:46,0,1,1
BRF:2
BRH:2
DA:14,1
DA:15,1
DA:16,1
DA:17,1
DA:21,1
DA:33,1
DA:34,1
DA:35,1
DA:36,1
DA:41,1
DA:42,1
DA:43,1
DA:44,1
DA:45,1
DA:46,1
DA:47,1
DA:49,1
DA:58,1
DA:59,1
LH:19
LF:19
end_of_record
SF:utility/src/test/java/org/oppia/android/util/parser/math/MathModelTest.kt
FN:9,org/oppia/android/util/parser/math/MathModelTest::<init> ()V
FN:90,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_diffModelByLineHeight_outsideTwoDecimals_returnsDiffKeyWithDiffDigest ()V
FN:69,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_diffModelByLineHeight_withinTwoDecimals_returnsSameKeyWithSameDigest ()V
FN:109,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_differentModelByInlineRendering_returnsDifferentKeyWithDifferentDigest ()V
FN:31,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_differentModelByLatex_returnsDifferentKeyWithDifferentDigest ()V
FN:50,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_differentModelByLineHeight_returnsDifferentKeyWithDifferentDigest ()V
FN:13,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_sameModelByValues_returnsSameKeyWithSameDigest ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::<init> ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_diffModelByLineHeight_outsideTwoDecimals_returnsDiffKeyWithDiffDigest ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_diffModelByLineHeight_withinTwoDecimals_returnsSameKeyWithSameDigest ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_differentModelByInlineRendering_returnsDifferentKeyWithDifferentDigest ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_differentModelByLatex_returnsDifferentKeyWithDifferentDigest ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_differentModelByLineHeight_returnsDifferentKeyWithDifferentDigest ()V
FNDA:1,org/oppia/android/util/parser/math/MathModelTest::testToKeySignature_sameModelByValues_returnsSameKeyWithSameDigest ()V
FNF:7
FNH:7
DA:9,1
DA:10,1
DA:13,1
DA:14,1
DA:15,1
DA:16,1
DA:18,1
DA:19,1
DA:20,1
DA:21,1
DA:23,1
DA:24,1
DA:25,1
DA:26,1
DA:27,1
DA:31,1
DA:32,1
DA:33,1
DA:34,1
DA:36,1
DA:37,1
DA:38,1
DA:39,1
DA:42,1
DA:43,1
DA:44,1
DA:45,1
DA:46,1
DA:50,1
DA:51,1
DA:52,1
DA:53,1
DA:55,1
DA:56,1
DA:57,1
DA:58,1
DA:61,1
DA:62,1
DA:63,1
DA:64,1
DA:65,1
DA:69,1
DA:70,1
DA:71,1
DA:72,1
DA:73,1
DA:75,1
DA:76,1
DA:77,1
DA:78,1
DA:82,1
DA:83,1
DA:84,1
DA:85,1
DA:86,1
DA:90,1
DA:91,1
DA:92,1
DA:93,1
DA:95,1
DA:96,1
DA:97,1
DA:98,1
DA:101,1
DA:102,1
DA:103,1
DA:104,1
DA:105,1
DA:109,1
DA:110,1
DA:111,1
DA:112,1
DA:113,1
DA:115,1
DA:116,1
DA:117,1
DA:118,1
DA:121,1
DA:122,1
DA:123,1
DA:124,1
DA:125,1
LH:82
LF:82
end_of_record

----------------

SF:utility/src/main/java/org/oppia/android/util/logging/ConsoleLogger.kt
FN:131,org/oppia/android/util/logging/ConsoleLogger$logToFileInBackground$1::invokeSuspend (Ljava/lang/Object;)Ljava/lang/Object;
FN:114,org/oppia/android/util/logging/ConsoleLogger$writeInternal$1::invokeSuspend (Ljava/lang/Object;)Ljava/lang/Object;
FN:18,org/oppia/android/util/logging/ConsoleLogger::<init> (Landroid/content/Context;Lkotlinx/coroutines/CoroutineDispatcher;ZZLorg/oppia/android/util/logging/LogLevel;Lorg/oppia/android/util/locale/OppiaLocale$MachineLocale;)V
FN:48,org/oppia/android/util/logging/ConsoleLogger::d (Ljava/lang/String;Ljava/lang/String;)V
FN:53,org/oppia/android/util/logging/ConsoleLogger::d (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FN:78,org/oppia/android/util/logging/ConsoleLogger::e (Ljava/lang/String;Ljava/lang/String;)V
FN:83,org/oppia/android/util/logging/ConsoleLogger::e (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FN:34,org/oppia/android/util/logging/ConsoleLogger::getLogErrorMessagesFlow ()Lkotlinx/coroutines/flow/SharedFlow;
FN:58,org/oppia/android/util/logging/ConsoleLogger::i (Ljava/lang/String;Ljava/lang/String;)V
FN:63,org/oppia/android/util/logging/ConsoleLogger::i (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FN:87,org/oppia/android/util/logging/ConsoleLogger::isLogEnable (Lorg/oppia/android/util/logging/LogLevel;)Z
FN:131,org/oppia/android/util/logging/ConsoleLogger::logToFileInBackground (Ljava/lang/String;)V
FN:38,org/oppia/android/util/logging/ConsoleLogger::v (Ljava/lang/String;Ljava/lang/String;)V
FN:43,org/oppia/android/util/logging/ConsoleLogger::v (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FN:68,org/oppia/android/util/logging/ConsoleLogger::w (Ljava/lang/String;Ljava/lang/String;)V
FN:73,org/oppia/android/util/logging/ConsoleLogger::w (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FN:95,org/oppia/android/util/logging/ConsoleLogger::writeError (Lorg/oppia/android/util/logging/LogLevel;Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FN:99,org/oppia/android/util/logging/ConsoleLogger::writeInternal (Lorg/oppia/android/util/logging/LogLevel;Ljava/lang/String;Ljava/lang/String;)V
FN:91,org/oppia/android/util/logging/ConsoleLogger::writeLog (Lorg/oppia/android/util/logging/LogLevel;Ljava/lang/String;Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger$logToFileInBackground$1::invokeSuspend (Ljava/lang/Object;)Ljava/lang/Object;
FNDA:1,org/oppia/android/util/logging/ConsoleLogger$writeInternal$1::invokeSuspend (Ljava/lang/Object;)Ljava/lang/Object;
FNDA:1,org/oppia/android/util/logging/ConsoleLogger::<init> (Landroid/content/Context;Lkotlinx/coroutines/CoroutineDispatcher;ZZLorg/oppia/android/util/logging/LogLevel;Lorg/oppia/android/util/locale/OppiaLocale$MachineLocale;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::d (Ljava/lang/String;Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::d (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FNDA:1,org/oppia/android/util/logging/ConsoleLogger::e (Ljava/lang/String;Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::e (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FNDA:1,org/oppia/android/util/logging/ConsoleLogger::getLogErrorMessagesFlow ()Lkotlinx/coroutines/flow/SharedFlow;
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::i (Ljava/lang/String;Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::i (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FNDA:1,org/oppia/android/util/logging/ConsoleLogger::isLogEnable (Lorg/oppia/android/util/logging/LogLevel;)Z
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::logToFileInBackground (Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::v (Ljava/lang/String;Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::v (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::w (Ljava/lang/String;Ljava/lang/String;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::w (Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FNDA:0,org/oppia/android/util/logging/ConsoleLogger::writeError (Lorg/oppia/android/util/logging/LogLevel;Ljava/lang/String;Ljava/lang/String;Ljava/lang/Throwable;)V
FNDA:1,org/oppia/android/util/logging/ConsoleLogger::writeInternal (Lorg/oppia/android/util/logging/LogLevel;Ljava/lang/String;Ljava/lang/String;)V
FNDA:1,org/oppia/android/util/logging/ConsoleLogger::writeLog (Lorg/oppia/android/util/logging/LogLevel;Ljava/lang/String;Ljava/lang/String;)V
FNF:19
FNH:7
BRDA:87,0,0,1
BRDA:87,0,1,0
BRDA:99,0,0,0
BRDA:99,0,1,1
BRDA:102,0,0,0
BRDA:102,0,1,1
BRDA:105,0,0,1
BRDA:105,0,1,0
BRDA:114,0,0,0
BRDA:114,0,1,1
BRF:10
BRH:5
DA:18,1
DA:19,1
DA:21,1
DA:22,1
DA:23,1
DA:24,1
DA:25,1
DA:27,1
DA:28,1
DA:30,1
DA:34,1
DA:38,0
DA:39,0
DA:43,0
DA:44,0
DA:48,0
DA:49,0
DA:53,0
DA:54,0
DA:58,0
DA:59,0
DA:63,0
DA:64,0
DA:68,0
DA:69,0
DA:73,0
DA:74,0
DA:78,1
DA:79,1
DA:83,0
DA:84,0
DA:87,1
DA:91,1
DA:92,1
DA:95,0
DA:96,0
DA:99,1
DA:100,0
DA:102,1
DA:103,1
DA:105,1
DA:106,0
DA:107,0
DA:112,1
DA:114,1
DA:115,1
DA:116,1
DA:117,1
DA:118,1
DA:119,1
DA:120,1
DA:123,1
DA:124,1
DA:131,0
DA:132,0
LH:30
LF:55
end_of_record
```
</details>

Which raises a doubt if the execution counts were just set to be **0 / 1**. 
- But the [geninfo docs](https://manpages.debian.org/stretch/lcov/geninfo.1.en.html#FILES) mention it as `DA:<line number>,<execution count>`.
- The [Bazel Coverage video](https://youtu.be/P51Rgcbxhyk?si=csg3lOziPllDVDFJ&t=461) too mentions it to represent the execution count.

But this [discussion](https://github.com/bazelbuild/rules_go/issues/3166#issuecomment-1141902942) in rules_go seems relavant to the exhibiting behaviour.

So I am just not sure right now if the represented data is represented right and also if we could add cases clearly distinguishing for lines hit multiple times, if the representation is done just with 0 and 1. Would it be possible to confirm if the data representation is handled or generated right over here?

Reply:
Ah, sorry what I meant was: for something like StateFragment that has both a StateFragmentTest and StateFragmentLocalTest. If they both have tests that cover the same line independently, how do we handle that here when merging the results into a single CoverageReport for StateFragment?

I don't think we care if a single line is hit more than once for a single test target since hit/not hit is the most, do we?

Reply:
Oh okay, the handling of overlapping coverage results is managed within the **`calculateAggregateCoverageReport`** function. Specifically, the function groups covered lines by their line numbers. If a line is covered by multiple test files, each test’s coverage for that line will be grouped together.

For instance, with the [tested sample data](https://github.com/oppia/oppia-android/pull/5459#issuecomment-2253028530), as both tests hit line number 8, the function groups them as follows:

```kotlin
Grouped lines: 
{
8=[
  line_number: 8 coverage: FULL, 
  line_number: 8 coverage: FULL
 ]
}
```

The **`aggregateCoverage`** function then determines the overall coverage for each line by checking if any entry in the group has **Coverage.FULL**. If so, the line’s coverage is set to **FULL**; otherwise, it is set to **NONE**. In this example, since both entries are **FULL**, the aggregated coverage for line **number 8** is **FULL**."

And to address the original comment: **"Yes"**, the existing 
- testRunCoverage_sharedAndLocalTestsMarkdownFormat_returnsCoverageData
- testRunCoverage_sharedAndLocalTestsHtmlFormat_returnsCoverageData

test cases cover this scenario. Both test cases independently hit the same line `a + b`.

Did the explanation and test cases address the scenario appropriately, or would it require any additional inclusion?

Reply:
Isn't it the case that the returned CoverageReport here is actually performing the merging, though? Since it would presumably return a CoverageReport for both test targets but one source file? It would be nice to verify the functionality of multiple targets on the same line for this function vs. later on in the report generation.

Reply:
As discussed, there will be a separate test added in a future PR once the coverage report itself can be verified.